### PR TITLE
[EPIC] Usability changes for multiple cursors functionality

### PR DESCRIPTION
**Pull request by sdwolfz**
Improvements made to multiple cursors support:

- New keybindings <kbd>C-M-j</kbd> and <kbd>C-M-k</kbd> for easier cursor creation while editing (inspired from sublime)
- Functionality extracted into a separate layer, installed by default in the `.spacemacs.template`
- Allow multiple backends (currently `evil-mc` is implemented).

Planning next:
- A transient state to help with complex operations.
- Create cursors from visual selection
- Quickly restore cursors once canceled

This is a WIP, many more goodies to come...

Changes already cherry-picked:
- b2fbd7c207f60848ed06eab56254857ab3266044 - Encapsulates multiple cursors functionality in a layer
- ad6ab57564abc8c7e1be6a5397d099a907d9894d - Renames the multiple cursors paste functions
- f94b51d4e82d4522a6795e7ec801c97bb651a639 - Documentation for multiple cursors
- c5a5ee305ba621a0299f5852a97f93faf54b4377 - Document sublime inspired keybindings

### CODE REVIEW

File: layers/+misc/multiple-cursors/funcs.el [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
+;;
+;; Author: Codruț Constantin Gușoi <codrut.gusoi@gmail.com>
+;; URL: https://github.com/syl20bnr/spacemacs
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+(defun spacemacs//paste-transient-state-p ()
```
Comment (syl20bnr) [inline layers/+misc/multiple-cursors/funcs.el]: I would rename this function to `spacemacs//evil-mc-paste-transient-state-p`. The current name is too broad.

**Comment by syl20bnr** [comment]
Great initiative!

Do you think we can achieve to lazy-load it ?

I see that you prepared the ground for supporting several multiple cursor backends. I'm not sure this is really possible to support Magnar's multiple cursors or if it even makes sense to do it.
Spacemacs needs evil even for Emacs style. If users choose the barebone Spacemacs without Evil then they can install Magnar's MC themselves.

We can work to bring all the goodies (essentially some handy key bindings) from Magnar's MC to our layer based on `evil-mc`.

On my side I'll try to make `evil-escape` work with it, if I cannot make it then we will have to disable it as we do with the paste transient state.

**Comment by sdwolfz** [comment]
I am not yet sure about removing backends and focusing only on `evil-mc`, but I requested feedback from multiple people before starting this and there seemed to be a lot of demand for having that.

It's going to take me a lot of time to finish this, so I will start integrating small changes directly and use this PR as an epic to track my progress.

**Comment by robbyoconnor** [comment]
I wanna learn how to use this so bad

**Comment by sdwolfz** [comment]
@robbyoconnor You can use the current functionality by enabling evil-mc with: 
```
(global-evil-mc-mode 1)
``` 
Then you have all the keybindings under `g r` (no `SPC`, just those two).

What I've done so far was to add 2 more keybindings, the rest is just moving code from one place to another. Hopefully I can put some time aside this weekend to add more stuff.

File: layers/+misc/multiple-cursors/funcs.el [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
+;;
+;; Author: Codruț Constantin Gușoi <codrut.gusoi@gmail.com>
+;; URL: https://github.com/syl20bnr/spacemacs
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+(defun spacemacs//evil-mc-paste-transient-state-p ()
```
Comment (am) [inline layers/+misc/multiple-cursors/funcs.el]: Is the double slash `//` a typo?

Comment (sdwolfz) [reply layers/+misc/multiple-cursors/funcs.el]: It is equivalent to `--` which denotes a function is "private" and should not be used in other files, or relied upon to have a stable API, or still exist in the future.

**Comment by bet4it** [comment]
Argument of `interactive` in `spacemacs/evil-mc-paste-before` and `spacemacs/evil-mc-paste-after` is wrong.
More details in #11340.

**Comment by sdwolfz** [comment]
@duianto @bet4it I saw your comments and PRs and will deal with the issues soon.

@deb0ch The commits are already cherry-picked into develop, see the main post for a refference to them. I want to keep this PR active to more easily test out stuff, that's why you can only see the placeholder commit.

**Comment by deb0ch** [comment]
@sdwolfz Omg, I actually tried to look in develop if they were not there but somehow managed to not see them 😅 Wonder what went on in my head

**Comment by sdwolfz** [comment]
Closing this as I'm not actively working on it, and I won't be able to in the near future.

**Pull request closed without merging**